pyclibrary/c_model.py

```diff
@@ -0,0 +1,589 @@
+import itertools
+from future.moves.itertools import zip_longest
+
+
+def _lpadded_str(text):
+    """An internal helper, returns '' if text is None otherwise ' '+text"""
+    if text is None:
+        return ''
+    else:
+        return ' ' + text
+
+
+class UnknownCustomType(KeyError):
+    """Thrown when a CLibType could not be resolved"""
+
+
+class CLibBase(object):
+
+    __slots__ = []
+
+    def _getattrnames(self):
+        """Internal method that lists all fields
+        :rtype: list[str]
+        """
+        for cls in type(self).__mro__:
+            if hasattr(cls, '__slots__'):
+                for name in cls.__slots__:
+                    yield name
+
+    def copy(self):
+        """Creates a shallow copy o this type.
+        :rtype: CLibBase
+        """
+        attrs = {anm: getattr(self, anm) for anm in self._getattrnames()}
+        return type(self)(**attrs)
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        return all(getattr(self, name) == getattr(other, name)
+                   for cls in type(self).__mro__
+                   for name in getattr(cls, '__slots__', ()))
+
+    def __ne__(self, other):
+        return not self == other
+
+    # noinspection PyUnresolvedReferences
+    def __repr__(self):
+        code_obj = self.__init__.__code__
+        init_params = code_obj.co_varnames[1:code_obj.co_argcount]
+        default_vals = dict(zip(reversed(init_params),
+                                reversed(self.__init__.__defaults__ or [])))
+
+        def par_repr(param_name):
+            val = getattr(self, param_name)
+            if param_name not in default_vals:
+                yield repr(val)
+            else:
+                default_val = ([] if param_name == "quals"
+                               else default_vals[param_name])
+                if val != default_val:
+                    yield param_name + '=' + repr(val)
+
+        return (type(self).__name__ + '(' +
+                ', '.join(itertools.chain(*map(par_repr, init_params))) + ')')
+
+
+class CLibType(CLibBase):
+    """This is the abstract base class for all objects that are modeling a
+    "C" type.
+    """
+
+    __slots__ = ['quals']
+
+    def __init__(self, quals=None):
+        """
+        :param list[str]|None quals: a list of type qualifiers
+            attached to this type.
+        """
+        self.quals = quals or []
+
+    def resolve(self, typedefs):
+        """Resolves all references to type definitions. A reference to
+        a type definition is represented by the TypeRef object. All
+        CustomType objects are resolved with the help of typedefs.
+
+        :param dict[str, CLibType] typedefs: maps the name of all known
+            to the corresponding CLibType.
+        :returns: a CLibType descendant object, that contains no TypeRef's.
+        :rtype: CLibType
+        """
+        return self
+
+    def c_repr(self, referrer_c_repr=None):
```

Comment:
As you are the maintainer of the project it is your decision how comments look like.
But I would ask you to think at least about the big advantage of markup syntax like sphinx to the numpy convention: It is machine readable!

Thus advanced developer tools (i.e. the widespread PyCharm) can use this information to provide powerful services like 
 * code completion
 * automatic refactoring
 * show warnings on TypeErrors, that otherwise would pop up at runtime.